### PR TITLE
[Spark] Handle case when Checkpoints.findLastCompleteCheckpoint is passed MAX_VALUE

### PR DESCRIPTION
**Pull request by prakharjain09**
#### Which Delta project/connector is this regarding?

- [X] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

## Description

Fixes an issue where `Checkpoints.findLastCompleteCheckpoint` goes into an almost infinite loop if it is passed a Checkpoint.MAX_VALUE.

## How was this patch tested?

UT

## Does this PR introduce _any_ user-facing changes?

No

### CODE REVIEW

File: spark/src/main/scala/org/apache/spark/sql/delta/Checkpoints.scala [hunk 1]
```diff
@@ -429,10 +429,46 @@ trait Checkpoints extends DeltaLogging {
...
+
+  private def findLastCompleteCheckpointBeforeInternal(
+      eventData: mutable.Map[String, String],
+      checkpointInstance: Option[CheckpointInstance]): Option[CheckpointInstance] = {
+    val upperBoundCv =
+      checkpointInstance
+        // If someone passes the upperBound as 0 or sentinel value, we should not do backward
+        // listing. Instead we should list the entire directory from 0 and return the latest
+        // available checkpoint.
+        .filterNot(cv => cv.version < 0 || cv.version == CheckpointInstance.MaxValue.version)
```
Comment (prakharjain09) [inline spark/src/main/scala/org/apache/spark/sql/delta/Checkpoints.scala:461]: This is the fix and there is a corresponding test: `findLastCompleteCheckpoint with CheckpointInstance.MAX value` .

File: spark/src/test/scala/org/apache/spark/sql/delta/FindLastCompleteCheckpointSuite.scala [hunk 2]
```diff
@@ -117,11 +139,15 @@ class FindLastCompleteCheckpointSuite
...
-          .contains(CheckpointInstance(version = 1000)))
+      val eventData1 = getLastCompleteCheckpointUsageLog {
+        assert(
+          log.findLastCompleteCheckpointBefore(Some(CheckpointInstance(version = 2000)))
+            .contains(CheckpointInstance(version = 1000)))
+      }
       assert(logStore.listFromCount == 1)
       assert(logStore.elementsConsumedFromListFromIter == 1002 + 2) // commits + checkpoint
+      assert(eventData1("iterations") == "1")
+      assert(eventData1("numFilesScanned") == "1004")
```
Comment (scovich) [inline spark/src/test/scala/org/apache/spark/sql/delta/FindLastCompleteCheckpointSuite.scala:150]: Does this work for better error messages?
```suggestion
      assert(eventData1("iterations") === "1")
      assert(eventData1("numFilesScanned") === "1004")
```

(bunch of others, including existing code... maybe worth a cleanup PR some day)

**Review by scovich — APPROVED** [review]
Nice catch

**Review by tdas — APPROVED** [review]

File: spark/src/test/scala/org/apache/spark/sql/delta/FindLastCompleteCheckpointSuite.scala (hunk 2 above)
Comment (prakharjain09) [reply spark/src/test/scala/org/apache/spark/sql/delta/FindLastCompleteCheckpointSuite.scala:150]: For consistency in the test suite, I am keeping these as `==`. Will open a PR which converts all `==` to `===`.

**Pull request merged**